api/services/ElasticSearchService.js

```diff
@@ -107,7 +133,49 @@ module.exports = {
     return RedisService.getJSONFromCache('percentiles', res, RedisService.DAILY, function() {
       return ElasticSearchService.lastMonthPercentiles();
     });
-  }
-
-
+  },
+  //download first 10000 results and proceed by processing and scrolling
+  lastMonthDownloadsBulk:function(days,callback){
+      hits = [];
+      var body = ElasticSearchService.queries.filters.lastMonthDownloads(days);
+      return es.search({
+      scroll:'5M',
+      index: 'stats',
+      body: body,
+      },function processAndGetMore(error,response){
+            //check the response
+        if (typeof response == "undefined" || typeof response.hits == "undefined") {    
+          console.log("you received an undefined response, response:"+response);
+          console.log("this was probably caused because there were no stats yet for this day");
+          console.log("or processing time took over 5 minutes (the scroll interval");
+        callback();
+        }  
+        DownloadStatsService.processDownloads(response,{},{},10000,callback);
+        });
+  },
+  //scroll further in search result, when response already contains a scroll id
+  scrollLastMonthDownloadsBulk:function(response,date,directDownloads,indirectDownloads,total,callback){
+    console.log("processing next 10000 records");
+    if (response.hits.total > total) {
+        // now we can call scroll over and over
+        es.scroll({
+          scrollId: response._scroll_id,
+          scroll: '5M'
+        }, function processScroll(error,response){
+            if (typeof response == "undefined" || typeof response.hits == "undefined") {    
+              console.log("you received an undefined response, response:"+response);
+              console.log("this was probably caused because there were no stats yet for this day");
+              console.log("or processing time took over 5 minutes (the scroll interval");
+            callback();
+            }  
+            return DownloadStatsService.processDownloads(response,directDownloads,indirectDownloads,total+10000,callback);
+        });
+      } else {
+        //write the responses to the database when done
+          DownloadStatsService.writeSplittedDownloadCounts(date,directDownloads,indirectDownloads).then(function(){
+            callback();
```

Comment:
here you can send back the result of the write as a result, so that would become
```
.then(function(result) {
  callback(null, result);
});
```